source/s2n/tls_channel_handler.c

```diff
@@ -0,0 +1,738 @@
+        size_t protocols_size = 4;
+        if (s_parse_protocol_preferences(options->alpn_list, protocols_cpy, &protocols_size)) {
+            aws_raise_error(AWS_IO_TLS_CTX_ERROR);
+            goto cleanup_conn;
+        }
+
+        const char *protocols[4];
+        AWS_ZERO_ARRAY(protocols);
+        for (size_t i = 0; i < protocols_size; ++i) {
+            protocols[i] = protocols_cpy[i];
+        }
+
+        if (s2n_connection_set_protocol_preferences(
+                s2n_handler->connection, (const char *const *)protocols, (int)protocols_size)) {
+            aws_raise_error(AWS_IO_TLS_CTX_ERROR);
+            goto cleanup_conn;
+        }
+    }
+
+    if (s2n_connection_set_config(s2n_handler->connection, s2n_ctx->s2n_config)) {
+        aws_raise_error(AWS_IO_TLS_CTX_ERROR);
+        goto cleanup_conn;
+    }
+
+    return &s2n_handler->handler;
+
+cleanup_conn:
+    s2n_connection_free(s2n_handler->connection);
+
+cleanup_s2n_handler:
+    aws_mem_release(allocator, s2n_handler);
+
+    return NULL;
+}
+
+struct aws_channel_handler *aws_tls_client_handler_new(
+    struct aws_allocator *allocator,
+    struct aws_tls_ctx *ctx,
+    struct aws_tls_connection_options *options,
+    struct aws_channel_slot *slot) {
+
+    return s_new_tls_handler(allocator, ctx, options, slot, S2N_CLIENT);
+}
+
+struct aws_channel_handler *aws_tls_server_handler_new(
+    struct aws_allocator *allocator,
+    struct aws_tls_ctx *ctx,
+    struct aws_tls_connection_options *options,
+    struct aws_channel_slot *slot) {
+
+    return s_new_tls_handler(allocator, ctx, options, slot, S2N_SERVER);
+}
+
+void aws_tls_ctx_destroy(struct aws_tls_ctx *ctx) {
+    struct s2n_ctx *s2n_ctx = ctx->impl;
+
+    if (s2n_ctx) {
+        s2n_config_free(s2n_ctx->s2n_config);
+        aws_mem_release(ctx->alloc, s2n_ctx);
+    }
+}
+
+static struct aws_tls_ctx *s_tls_ctx_new(
+    struct aws_allocator *alloc,
+    struct aws_tls_ctx_options *options,
+    s2n_mode mode) {
+    struct s2n_ctx *s2n_ctx = (struct s2n_ctx *)aws_mem_acquire(alloc, sizeof(struct s2n_ctx));
+
+    if (!s2n_ctx) {
+        return NULL;
+    }
+
+    s2n_ctx->ctx.alloc = alloc;
+    s2n_ctx->ctx.impl = s2n_ctx;
+    s2n_ctx->s2n_config = s2n_config_new();
+
+    if (!s2n_ctx->s2n_config) {
+        goto cleanup_s2n_ctx;
+    }
+
+    switch (options->minimum_tls_version) {
+        case AWS_IO_SSLv3:
+            s2n_config_set_cipher_preferences(s2n_ctx->s2n_config, "CloudFront-SSL-v-3");
+            break;
+        case AWS_IO_TLSv1:
+            s2n_config_set_cipher_preferences(s2n_ctx->s2n_config, "CloudFront-TLS-1-0-2016");
+            break;
+        case AWS_IO_TLSv1_1:
+            s2n_config_set_cipher_preferences(s2n_ctx->s2n_config, "CloudFront-TLS-1-1-2016");
+            break;
+        case AWS_IO_TLSv1_2:
+            s2n_config_set_cipher_preferences(s2n_ctx->s2n_config, "CloudFront-TLS-1-2-2018");
+            break;
+        case AWS_IO_TLSv1_3:
```

Comment:
This is a failure on darwin, should it be a failure here? Is falling back to a version of TLS that the user isn't expecting an ok thing to do here?

Reply:
yeah, should be an error.